routers/api/packages/arch/arch.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"bytes"
+	"encoding/hex"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/modules/context"
+	arch_module "code.gitea.io/gitea/modules/packages/arch"
+	"code.gitea.io/gitea/routers/api/packages/helper"
+	arch_service "code.gitea.io/gitea/services/packages/arch"
+)
+
+// Push new package to arch package registry.
+func Push(ctx *context.Context) {
+	var (
+		owner    = ctx.Params("username")
+		filename = ctx.Req.Header.Get("filename")
+		distro   = ctx.Req.Header.Get("distro")
+		sign     = ctx.Req.Header.Get("sign")
+	)
+
+	// Read package to memory for package validation.
+	pkgdata, err := io.ReadAll(ctx.Req.Body)
```

Comment:
It's not save to read the full file into memory. Use `CreateHashedBufferFromReader`.